source/slang/ir-legalize-types.cpp

```diff
@@ -1293,7 +1313,7 @@ static LegalVal legalizeFunc(
         newResultType);
 
     context->builder->setDataType(irFunc, newFuncType);
-
+    context->builder->setInsertInto(irFunc);
```

Comment:
I don't know why it worked before, but it seems to me that we need to call `setInsertInto` here, or otherwise builder may be in an outdated state as a result of previous legalization operation.

Reply:
Yes, but this does not seem to be related to the new return logic (as it happens before it). The crash is at the first instruction (an IRVar that needs to be legalized). IRVar goes through a separate legalization step (`legalizeVar`) that is called before the `legalizeInstr` calling `setInsertBefore()`. I have no idea why it did not crash before...

Reply:
I found the problem. In ir-legalize-types.cpp:1058, we are setting `context->insertBeforeLocalVar`, which we use at line 1385. However, the new var to be inserted is created by emitVar, which calls addInst to add it to an undefined place, and then removed at line 1384.

We should modify emitVar, so that it does not add the new inst to some invalid location.